bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/internal/OpenThermGatewaySocketConnector.java

```diff
@@ -0,0 +1,210 @@
+    private static final int COMMAND_RESPONSE_TIME_MILLISECONDS = 100;
+    private static final int COMMAND_TIMEOUT_MILLISECONDS = 5000;
+
+    private OpenThermGatewayCallback callback;
+    private String ipaddress;
+    private int port;
+
+    private @Nullable PrintWriter writer;
+
+    private volatile boolean stopping;
+    private boolean connected;
+
+    private final Logger logger = LoggerFactory.getLogger(OpenThermGatewaySocketConnector.class);
+
+    private Map<String, Entry<Long, GatewayCommand>> pendingCommands = new HashMap<>();
+
+    public OpenThermGatewaySocketConnector(OpenThermGatewayCallback callback, String ipaddress, int port) {
+        this.callback = callback;
+        this.ipaddress = ipaddress;
+        this.port = port;
+    }
+
+    @Override
+    public void run() {
+        stopping = false;
+        connected = false;
+        
+        logger.info("Connecting OpenThermGatewaySocketConnector to {}:{}", this.ipaddress, this.port);
+
+        callback.connecting();
+        
+        try (Socket socket = new Socket()) {            
+            socket.connect(new InetSocketAddress(this.ipaddress, this.port), COMMAND_TIMEOUT_MILLISECONDS);
+            socket.setSoTimeout(COMMAND_TIMEOUT_MILLISECONDS);
+    
+            connected = true;
+    
+            callback.connected();
+    
+            logger.info("OpenThermGatewaySocketConnector connected");
+    
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                PrintWriter wrt = new PrintWriter(socket.getOutputStream(), true)) {
+                    // Make writer accessible on class level
+                    writer = wrt;
+    
+                    sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintReport, "A"));
+                    // Set the OTGW to report every message it receives and transmits
+                    sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintSummary, "0"));
+        
+                    while (!stopping && !Thread.currentThread().isInterrupted()) {
+                        @Nullable String message = reader.readLine();
+        
+                        if (message != null) {
+                            handleMessage(message);
+                        } else {
+                            logger.info("Connection closed by OpenTherm Gateway");
+                            break;
+                        }
+                    }
+        
+                    logger.info("Stopping OpenThermGatewaySocketConnector");
+            }
+            finally {
+                connected = false;
+    
+                logger.info("OpenThermGatewaySocketConnector disconnected");
+                callback.disconnected();
+            }
+        } catch (IOException ex) {
+            logger.warn("Unable to connect to the OpenTherm Gateway." , ex);
+        }        
+    }
+
+    @Override
+    public synchronized void stop() {
```

Comment:
why is this synchronized?

Reply:
Since i'm using callback methods and afaik dont have resources shared between threads.. it seems that synchronized isnt neccesary at all. Removed all occurences.